### PR TITLE
WIP: Using only ctypes to construct the ringbuffer

### PR DESCRIPTION
**Pull request by macat**
This PR converts the ringbuffer to rely only on ctypes objects.

- Cleaner API
- Removed SlotArray abstraction
- Read method returns an array instead of a single object, allowing faster, batch reads
- Started converting the code to use type hints
- Some code formatting with YAPF

### CODE REVIEW

File: ringbuffer.py [hunk 1]
```diff
@@ -93,15 +97,16 @@ class RingBuffer:
     will not work.
     """
 
-    def __init__(self, *, slot_bytes, slot_count):
+    def __init__(self, *, c_type: Type[ctypes._SimpleCData], slot_count: int):
```
Comment (alanhdu) [inline ringbuffer.py:100]: Hm... my suspicion is that we should be using a generic typevar with an upper bound (something like `T = TypeVar("T", bound=ctypes._SimpleCData)`.

File: ringbuffer.py [hunk 2]
```diff
@@ -185,23 +189,29 @@ def try_write(self, data):
             self.array[position.index] = data
             self.writer.increment()
 
-    def _has_read_conflict(self, reader_position):
+    def _try_read_no_lock(self, reader, length: int = 1):
```
Comment (alanhdu) [inline ringbuffer.py]: Do you mind typing `reader` too? I think it's `reader: Pointer`.

File: ringbuffer.py [hunk 3]
```diff
@@ -253,7 +263,7 @@ def force_reader_sync(self):
                 reader.set(writer_position.counter)
 
             for reader in self.readers:
-                p = reader.get()
+                reader.get()
```
Comment (alanhdu) [inline ringbuffer.py:274]: Is this loop even necessary? We can wait to synchronize `reader.position` later, right?

(also, do you know why we have the `counter` is a `ctypes.c_longlong` thing? Is it a performance optimization, or something?)

File: ringbuffer.py [hunk 4]
```diff
@@ -185,23 +189,29 @@ def try_write(self, data):
...
+        max_length = writer_position.counter - position.counter
+
+        if max_length < length:
+            raise NotEnoughItemsError
+
+        new_array = (self.c_type * length)()
+        ctypes.memmove(new_array,
+                       ctypes.addressof(self.array[position.index]),
+                       ctypes.sizeof(new_array))
+        reader.increment_by(length)
```
Comment (alanhdu) [inline ringbuffer.py]: Nice!

Should we document somewhere that the `ctypes` struct shouldn't contain internal pointers that might be invalidated here though (or really pointers in general, because I think the semantics get wonky there)? I can't really imagine that we'd every use one of those, but it'd be good to note it just in case.

File: ringbuffer.py [hunk 5]
```diff
@@ -77,9 +78,12 @@ def get(self):
         self.position.counter = self.counter.value
         return self.position
 
-    def set(self, counter):
+    def set(self, counter: int):
         self.counter.value = counter
 
+    def increment_by(self, n: int):
```
Comment (alanhdu) [inline ringbuffer.py]: Nit: can you move this function next to `increment`?

**Review by alanhdu — APPROVED** [review]
One or two nits and tests are failing, but otherwise LGTM!

File: ringbuffer.py (hunk 3 above)
Comment (macat) [reply ringbuffer.py:274]: Not sure about the loop. I think the counter needs to work across process boundaries, thus a rawvalue is required.

File: ringbuffer.py (hunk 4 above)
Comment (macat) [reply ringbuffer.py]: Yes, I'm going to add docs about that.

File: ringbuffer.py [hunk 6]
```diff
@@ -185,23 +189,51 @@ def try_write(self, data):
...
+            # generation, a simple memmove can copy the array.
+            ctypes.memmove(new_array,
+                           ctypes.addressof(self.array[position.index]),
+                           ctypes.sizeof(new_array))
+        else:
+            # When the writer has a greater generation count, two memmove
+            # calls are necessary to fill the new array with data.
+            first_slice_length = remaining_buffer_length
+            second_slice_length = length - first_slice_length
+            one_item_size = int(ctypes.sizeof(new_array) / length)
```
Comment (csosborn) [inline ringbuffer.py]: Is this different from `ctypes.sizeof(self.c_type)`?

File: ringbuffer.py [hunk 7]
```diff
@@ -185,23 +189,51 @@ def try_write(self, data):
...
+
+            ctypes.memmove(new_array,
+                           ctypes.addressof(self.array[position.index]),
+                           first_slice_length * one_item_size)
+
+            if second_slice_length > 0:
```
Comment (csosborn) [inline ringbuffer.py]: Is this conditional necessary? Would changing `>` to `>=` on line 210 make it unnecessary?

File: ringbuffer.py [hunk 8]
```diff
@@ -220,9 +252,9 @@ def try_read(self, reader):
                 order to wait for new data to arrive.
         """
         with self.lock.for_read():
-            return self._try_read_no_lock(reader)
+            return self._try_read_no_lock(reader, length=length)
```
Comment (csosborn) [inline ringbuffer.py:241]: I believe the `Returns` line in the doc block above is now incorrect. This returns a ctypes array.

**Review by csosborn — APPROVED** [review]
This looks good. I left a few comments but nothing that should block this. I agree that the try_read method is complicated, but as long as the person reading it knows that it's trying to stitch together one buffer using data from both the end of one ring buffer cycle and the beginning of the next, then it's not so hard to follow. A quick comment to that effect will help.

File: ringbuffer.py (hunk 6 above)
Comment (macat) [reply ringbuffer.py]: good catch, simplifying this.

File: ringbuffer.py (hunk 7 above)
Comment (macat) [reply ringbuffer.py]: yes! thanks!

File: ringbuffer.py [hunk 9]
```diff
@@ -185,23 +189,52 @@ def try_write(self, data):
...
+
+            ctypes.memmove(new_array,
+                           ctypes.addressof(self.array[position.index]),
+                           first_slice_length * one_item_size)
+
+            if second_slice_length > 0:
+                ctypes.memmove(
+                    ctypes.addressof(new_array[first_slice_length]),
+                    ctypes.addressof(self.array[0]),
+                    second_slice_length * one_item_size)
```
Comment (alanhdu) [inline ringbuffer.py]: Nice catch! I think this logic can be simplified though:

```python
item_size = ctypes.sizeof(self.c_type)
remaining = self.slot_count - reader_position.index
ctypes.memmove(
    new_array,
    ctypes.addressof(self.array[reader_position.index]),
    item_size * min(length, remaining)
)
if length > remaining:
    ctypes.memmove(
        ctypes.addressof(new_array[remaining]),
        ctypes.addressof(self.array),
        (length - remaining) * item_size
    )

```

Comment (macat) [reply ringbuffer.py]: Actually I went with the recommendation from Chris just at the time when you posted this.

File: ringbuffer.py [hunk 10]
```diff
@@ -185,32 +190,57 @@ def try_write(self, data):
...
+            second_slice_length = length - first_slice_length
+
+            ctypes.memmove(new_array,
+                           ctypes.addressof(self.array[position.index]),
+                           first_slice_length * self.slot_size)
+
+            ctypes.memmove(
+                ctypes.addressof(new_array[first_slice_length]),
+                ctypes.addressof(self.array[0]),
+                second_slice_length * self.slot_size)
```
Comment (alanhdu) [inline ringbuffer.py]: Reposting because I think my comment got lost in Github:

I think that https://github.com/ctrl-labs/ringbuffer/blob/0c0550b06b0b5eb77694ec0a18e26b261c1760d2/ringbuffer.py#L211-L231

can just replaced with

```python
item_size = ctypes.sizeof(self.c_type)
remaining = self.slot_count - reader_position.index
ctypes.memmove(
    new_array,
    ctypes.addressof(self.array[reader_position.index]),
    item_size * min(length, remaining)
)
if length > remaining:
    ctypes.memmove(
        ctypes.addressof(new_array[remaining]),
        ctypes.addressof(self.array),
        (length - remaining) * item_size
    )
```

which I think is a little cleaner (unconditionally memmove the "first part" and only have the second memmove in a conditional) and makes it more obvious why this is necessary.

**Review by nishkakar — APPROVED** [review]
Lgtm!

File: ringbuffer.py (hunk 10 above)
Comment (macat) [reply ringbuffer.py]: @alanhdu awesome improvement! Thanks a lot!
